--- ideapad-laptop.c    2025-04-17 12:41:49.000000000 +0800
+++ ideapad-laptop-tb.c 2025-04-17 12:48:29.380615597 +0800
@@ -32,7 +32,7 @@
 #include <linux/sysfs.h>
 #include <linux/types.h>
 #include <linux/wmi.h>
-#include "ideapad-laptop.h"
+#include "ideapad-laptop-tb.h"

 #include <acpi/video.h>

@@ -265,7 +265,7 @@
 /*
  * ACPI Helpers
  */
-#define IDEAPAD_EC_TIMEOUT 200 /* in ms */
+#define IDEAPAD_EC_TIMEOUT 50 /* in ms */

 static int eval_int(acpi_handle handle, const char *name, unsigned long *res)
 {
